### ISSUE TITLE
Community fund proposal votingCycle does not stop incrementing when the state is terminal

### ISSUE THREAD

**Issue by dantudor** [NAVCoin/navcoin-core#391]
When a proposal state has changed to a terminal state the votingCycle should remain at the previous value and no longer increment for future cycles.

Currently when `navcoin-cli getproposal bdc588af828557deb114f67c5dfa87e44e4f25b142ac091aa6d65ee653cbf44a` is executed, a votingCycle value of 7.

This value should have remained at the votingCycle value of acceptance.

**Comment by marcus290** [comment]
by design, the votingCycle goes +1 over. `IsExpired()` relies on this to check for voting period expiry.

Otherwise, the votingCycle value shouldn't go any higher. I wasn't able to replicate this in my own 4.5.1 wallet. `getproposal bdc588af828557deb114f67c5dfa87e44e4f25b142ac091aa6d65ee653cbf44a` gave me a voting cycle of 5 not 7. Any other way to replicate this?

Nevertheless, I think we should stop counting those votes in votingCycle 7.

**Comment by aguycalled** [comment]
Yes your analysis is correct. With the current implementation the proposal can get votes in `nCyclesPaymentRequestVoting+1` but will never morph to an state different than `EXPIRED` so the votes have no effect.

Ideally the proposal should morph in a `EXPIRED WAITING FOR END OF VOTING PERIOD` state when on cycle `nCyclesPaymentRequestVoting` and morph to `EXPIRED` on `nCyclesPaymentRequestVoting+1` if it was not `ACCEPTED` or `REJECTED` on cycle `nCyclesPaymentRequestVoting`.

The fix would be to change the check in IsExpired from:

```nVotingCycle > Params().GetConsensus().nCyclesProposalVoting ```

to:

```nVotingCycle >= Params().GetConsensus().nCyclesProposalVoting ```

And to break the if-elseif branch in CountVotes() to

```
if(proposal.IsExpired(pindexNew->GetBlockTime())){
        ....
}
if(proposal.IsRejected()){
        ....
}
...
```

Example of code for my proposal of fix (missing the fork part) at https://github.com/aguycalled/navcoin-core/tree/cfund-last-cycle

**Comment by marcus290** [comment]
Before proposals/prequests can be accepted, proposals will check `(proposal.fState == CFund::NIL || proposal.fState == CFund::PENDING_FUNDS)` and payment requests will check `(prequest.fState == CFund::NIL)` .

When I was testing the proposed fix, a proposal/prequest that gets enough votes in the last votingCycle will not be accepted because fState is changed to `EXPIRED`  before the check and it will fail such check.

**Comment by aguycalled** [comment]
@marcus290 you are completely right with this, this is going to require a proper rethink of how the cfund works. not easy solution

**Comment by marcus290** [comment]
@aguycalled well I haven't checked it yet but https://github.com/aguycalled/navcoin-core/commit/9c1db6473f7457e09e893625e40968c728930c18 should fix the main issue of this thread. 

Then we could keep the existing design for now to increment +1 votingCycle after a state change. Craig suggested we could just prevent votes on `nCyclesPaymentRequestVoting + 1` just to avoid confusion and bloat. What do you think?

**Comment by aguycalled** [comment]
makes sense as a temporary solution as long as it does not introduce changes which would require a future fork to normalise the situation

**Comment by marcus290** [comment]
Patched with #396
